From 00a6f3998cf94aee8ea77f791ec5734a552d585d Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Mon, 20 Jan 2025 14:18:23 +0100
Subject: [PATCH 1561/1695] core: lib: add rse_comms

For stm32mp2xx platform configure in m33tdcid, rse comms is
used to transport secure request to m33 secure.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 1308458b970eba61a7242e5a00a245c13cab8104 core: lib: add rse_comms]
Change-Id: Ibbbad3953c9fb056301768685764f1ab0991676c
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/434429
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/core.mk                                  |   5 +
 core/lib/rse_comms/include/psa/client.h       | 117 +++++++++++
 core/lib/rse_comms/include/rse_comms.h        |  21 ++
 core/lib/rse_comms/src/rse_comms.c            | 192 ++++++++++++++++++
 core/lib/rse_comms/src/rse_comms_protocol.c   |  82 ++++++++
 core/lib/rse_comms/src/rse_comms_protocol.h   |  61 ++++++
 .../rse_comms/src/rse_comms_protocol_common.h |  34 ++++
 .../rse_comms/src/rse_comms_protocol_embed.c  |  83 ++++++++
 .../rse_comms/src/rse_comms_protocol_embed.h  |  45 ++++
 .../src/rse_comms_protocol_pointer_access.c   |  72 +++++++
 .../src/rse_comms_protocol_pointer_access.h   |  47 +++++
 core/lib/rse_comms/src/sub.mk                 |   3 +
 core/lib/rse_comms/sub.mk                     |   2 +
 13 files changed, 764 insertions(+)
 create mode 100644 core/lib/rse_comms/include/psa/client.h
 create mode 100644 core/lib/rse_comms/include/rse_comms.h
 create mode 100644 core/lib/rse_comms/src/rse_comms.c
 create mode 100644 core/lib/rse_comms/src/rse_comms_protocol.c
 create mode 100644 core/lib/rse_comms/src/rse_comms_protocol.h
 create mode 100644 core/lib/rse_comms/src/rse_comms_protocol_common.h
 create mode 100644 core/lib/rse_comms/src/rse_comms_protocol_embed.c
 create mode 100644 core/lib/rse_comms/src/rse_comms_protocol_embed.h
 create mode 100644 core/lib/rse_comms/src/rse_comms_protocol_pointer_access.c
 create mode 100644 core/lib/rse_comms/src/rse_comms_protocol_pointer_access.h
 create mode 100644 core/lib/rse_comms/src/sub.mk
 create mode 100644 core/lib/rse_comms/sub.mk

diff --git a/core/core.mk b/core/core.mk
index b4690cae3..ab76601bd 100644
--- a/core/core.mk
+++ b/core/core.mk
@@ -185,6 +185,11 @@ libdir = core/lib/psa-adac
 include mk/lib.mk
 endif
 
+ifeq ($(CFG_RSE_COMMS),y)
+libname = rse_comms
+libdir = core/lib/rse_comms
+include mk/lib.mk
+endif
 #
 # Do main source
 #
diff --git a/core/lib/rse_comms/include/psa/client.h b/core/lib/rse_comms/include/psa/client.h
new file mode 100644
index 000000000..7d954ab70
--- /dev/null
+++ b/core/lib/rse_comms/include/psa/client.h
@@ -0,0 +1,117 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2018-2024, Arm Limited. All rights reserved.
+ *
+ */
+
+#ifndef PSA_CLIENT_H
+#define PSA_CLIENT_H
+#include <psa/crypto_types.h>
+#include <psa/crypto_values.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <util.h>
+
+#ifndef IOVEC_LEN
+#define IOVEC_LEN(arr) ((uint32_t)ARRAY_SIZE(arr))
+#endif
+
+/*********************** PSA Client Macros and Types *************************/
+
+/**
+ * The version of the PSA Framework API that is being used to build the calling
+ * firmware. Only part of features of FF-M v1.1 have been implemented. FF-M v1.1
+ * is compatible with v1.0.
+ */
+#define PSA_FRAMEWORK_VERSION	(0x0101u)
+
+/**
+ * Return value from psa_version() if the requested RoT Service is not present
+ * in the system.
+ */
+#define PSA_VERSION_NONE	(0u)
+
+/**
+ * The zero-value null handle can be assigned to variables used in clients and
+ * RoT Services, indicating that there is no current connection or message.
+ */
+#define PSA_NULL_HANDLE		((psa_handle_t)0)
+
+/**
+ * Tests whether a handle value returned by psa_connect() is valid.
+ */
+#define PSA_HANDLE_IS_VALID(handle)	((psa_handle_t)(handle) > 0)
+
+/**
+ * Converts the handle value returned from a failed call psa_connect() into
+ * an error code.
+ */
+#define PSA_HANDLE_TO_ERROR(handle)	((psa_status_t)(handle))
+
+/**
+ * Maximum number of input and output vectors for a request to psa_call().
+ */
+#define PSA_MAX_IOVEC		(4u)
+
+/**
+ * The minimum and maximum value that can be passed
+ * as the type parameter in a call to psa_call().
+ */
+#define PSA_CALL_TYPE_MIN	(0)
+#define PSA_CALL_TYPE_MAX	(INT16_MAX)
+
+/**
+ * An IPC message type that indicates a generic client request.
+ */
+#define PSA_IPC_CALL		(0)
+typedef int32_t psa_handle_t;
+
+/**
+ * A read-only input memory region provided to an RoT Service.
+ */
+typedef struct psa_invec {
+	const void *base;	/*!< the start address of the memory buffer */
+	size_t len;		/*!< the size in bytes                      */
+} psa_invec;
+
+/**
+ * A writable output memory region provided to an RoT Service.
+ */
+typedef struct psa_outvec {
+	void *base;		/*!< the start address of the memory buffer */
+	size_t len;		/*!< the size in bytes                      */
+} psa_outvec;
+
+/**
+ * Call an RoT Service on an established connection.
+ *
+ * handle	A handle to an established connection.
+ * type		The request type. Must be zero(PSA_IPC_CALL) or positive.
+ * in_vec	Array of input psa_invec structures.
+ * in_len	Number of input psa_invec structures.
+ * out_vec	Array of output psa_outvec structures.
+ * out_len	Number of output psa_outvec structures.
+ *
+ * Return value >=0	RoT Service-specific status value.
+ * Return value <0	RoT Service-specific error code.
+ *
+ * PSA_ERROR_PROGRAMMER_ERROR:
+ *	- The connection has been terminated by the RoT Service.
+ *
+ * The call is a PROGRAMMER ERROR if one or more of the following are true:
+ *	- An invalid handle was passed.
+ *	- The connection is already handling a request.
+ *	- type < 0.
+ *	- An invalid memory reference was provided.
+ *	- in_len + out_len > PSA_MAX_IOVEC.
+ *	- The message is unrecognized by the RoT.
+ *	- Service or incorrectly formatted.
+ */
+psa_status_t psa_call(psa_handle_t handle,
+		      int32_t type,
+		      const psa_invec *in_vec,
+		      size_t in_len,
+		      psa_outvec *out_vec,
+		      size_t out_len);
+
+#endif /* PSA_CLIENT_H */
diff --git a/core/lib/rse_comms/include/rse_comms.h b/core/lib/rse_comms/include/rse_comms.h
new file mode 100644
index 000000000..49010f636
--- /dev/null
+++ b/core/lib/rse_comms/include/rse_comms.h
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022-2024, Arm Limited and Contributors. All rights reserved.
+ * Copyright (C) 2025, STMicroelectronics - All Rights Reserved
+ *
+ */
+
+#ifndef RSE_COMMS_H
+#define RSE_COMMS_H
+
+#include <stdint.h>
+
+typedef int (*rse_send_callback_t)(const uint8_t *send_buffer, size_t size);
+typedef int (*rse_recv_callback_t)(const uint8_t *receive_buffer, size_t *size);
+typedef int (*rse_get_max_message_size_callback_t)(size_t *size);
+
+int rse_register_cb(rse_send_callback_t send_cb, rse_recv_callback_t rcv_cb,
+		    rse_get_max_message_size_callback_t size_cb);
+void rse_set_client_id(uint16_t client_id);
+
+#endif /* RSE_COMMS_H */
diff --git a/core/lib/rse_comms/src/rse_comms.c b/core/lib/rse_comms/src/rse_comms.c
new file mode 100644
index 000000000..4bb759570
--- /dev/null
+++ b/core/lib/rse_comms/src/rse_comms.c
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2022-2024, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2025 STMicroelectronics  All rights reserved.
+ *
+ */
+
+#include <util.h>
+#include <kernel/panic.h>
+#include <psa/client.h>
+#include <rse_comms.h>
+#include "rse_comms_protocol.h"
+#include <stdint.h>
+#include <string.h>
+#include <trace.h>
+/*
+ * Union as message space and reply space are never used at the same time,
+ * and this saves space aswe can overlap them.
+ */
+union __packed __aligned(4) rse_comms_io_buffer_t {
+	struct serialized_rse_comms_msg_t msg;
+	struct serialized_rse_comms_reply_t reply;
+};
+
+static rse_get_max_message_size_callback_t rse_mbx_size_cb;
+static rse_send_callback_t rse_mbx_send_cb;
+static rse_recv_callback_t rse_mbx_recv_cb;
+static uint16_t rse_client_id = ~0;
+
+static inline int rse_mbx_get_max_message_size(size_t *size)
+{
+	return rse_mbx_size_cb ? rse_mbx_size_cb(size) : -1;
+}
+
+static inline int rse_mbx_send_data(const uint8_t *send_buffer, size_t size)
+{
+	return rse_mbx_send_cb ? rse_mbx_send_cb(send_buffer, size) : -1;
+}
+
+static inline int rse_mbx_receive_data(uint8_t *receive_buffer, size_t *size)
+{
+	return rse_mbx_recv_cb ? rse_mbx_recv_cb(receive_buffer, size) : -1;
+}
+
+static uint8_t select_protocol_version(const psa_invec *in_vec, size_t in_len,
+				       const psa_outvec *out_vec,
+				       size_t out_len)
+{
+	size_t comms_mbx_msg_size;
+	size_t comms_embed_msg_min_size;
+	size_t comms_embed_reply_min_size;
+	size_t in_size_total = 0;
+	size_t out_size_total = 0;
+	size_t i;
+
+	for (i = 0U; i < in_len; ++i)
+		in_size_total += in_vec[i].len;
+
+	for (i = 0U; i < out_len; ++i)
+		out_size_total += out_vec[i].len;
+
+	if (rse_mbx_get_max_message_size(&comms_mbx_msg_size))
+		panic();
+
+	comms_embed_msg_min_size = sizeof(struct serialized_rse_comms_header_t)
+		+ sizeof(struct rse_embed_msg_t)
+		- PLAT_RSE_COMMS_PAYLOAD_MAX_SIZE;
+
+	comms_embed_reply_min_size =
+		sizeof(struct serialized_rse_comms_header_t)
+		+ sizeof(struct rse_embed_reply_t)
+		- PLAT_RSE_COMMS_PAYLOAD_MAX_SIZE;
+
+	/* Use embed if we can pack into one message and reply, else use
+	 * pointer_access. The underlying mailbox transport protocol uses a
+	 * single uint32_t to track the length, so the amount of data that
+	 * can be in a message is 4 bytes less than rse_mbx_get_max_message_size
+	 * reports.
+	 *
+	 * TODO tune this with real performance numbers, it's possible a
+	 * pointer_access message is less performant than multiple embed
+	 * messages due to ATU configuration costs to allow access to the
+	 * pointers.
+	 */
+	if ((comms_embed_msg_min_size + in_size_total >
+	     comms_mbx_msg_size - sizeof(uint32_t)) ||
+	    (comms_embed_reply_min_size + out_size_total >
+	     comms_mbx_msg_size - sizeof(uint32_t)))
+		return RSE_COMMS_PROTOCOL_POINTER_ACCESS;
+	else
+		return RSE_COMMS_PROTOCOL_EMBED;
+}
+
+void rse_set_client_id(uint16_t client_id)
+{
+	rse_client_id = client_id;
+}
+
+int rse_register_cb(rse_send_callback_t send_cb, rse_recv_callback_t rcv_cb,
+		    rse_get_max_message_size_callback_t size_cb)
+{
+	rse_mbx_recv_cb = rcv_cb;
+	rse_mbx_send_cb = send_cb;
+	rse_mbx_size_cb = size_cb;
+
+	return 0;
+}
+
+psa_status_t psa_call(psa_handle_t handle, int32_t type,
+		      const psa_invec *in_vec, size_t in_len,
+		      psa_outvec *out_vec, size_t out_len)
+{
+	/*
+	 * Declared statically to avoid using huge amounts of stack space.
+	 * Maybe revisit if functions not being reentrant becomes a problem.
+	 */
+	static union rse_comms_io_buffer_t io_buf;
+	int err;
+	psa_status_t status;
+	static uint8_t seq_num = 1U;
+	size_t msg_size;
+	size_t reply_size = sizeof(io_buf.reply);
+	psa_status_t return_val;
+	size_t idx;
+
+	if (type > PSA_CALL_TYPE_MAX || type < PSA_CALL_TYPE_MIN ||
+	    in_len > PSA_MAX_IOVEC   || out_len > PSA_MAX_IOVEC)
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	io_buf.msg.header.seq_num = seq_num;
+	/*
+	 * No need to distinguish callers (currently concurrent calls
+	 * are not supported).
+	 */
+	io_buf.msg.header.client_id =  rse_client_id;
+	io_buf.msg.header.protocol_ver =
+		select_protocol_version(in_vec, in_len, out_vec,
+					out_len);
+
+	status = rse_protocol_serialize_msg(handle, type, in_vec, in_len,
+					    out_vec, out_len, &io_buf.msg,
+					    &msg_size);
+	if (status != PSA_SUCCESS)
+		return status;
+
+	FMSG("[RSE-COMMS] Sending message\n");
+	FMSG("protocol_ver=%u\n", io_buf.msg.header.protocol_ver);
+	FMSG("seq_num=%u\n", io_buf.msg.header.seq_num);
+	FMSG("client_id=%u\n", io_buf.msg.header.client_id);
+	for (idx = 0; idx < in_len; idx++) {
+		FMSG("in_vec[%lu].len=%lu\n", idx, in_vec[idx].len);
+		FMSG("in_vec[%lu].buf=%p\n", idx, (void *)in_vec[idx].base);
+	}
+
+	err = rse_mbx_send_data((uint8_t *)&io_buf.msg, msg_size);
+	if (err != 0)
+		return PSA_ERROR_COMMUNICATION_FAILURE;
+
+	/*
+	 * Poisoning the message buffer (with a known pattern).
+	 * Helps in detecting hypothetical RSE communication bugs.
+	 */
+	memset(&io_buf.msg, 0xA5, msg_size);
+
+	err = rse_mbx_receive_data((uint8_t *)&io_buf.reply, &reply_size);
+	if (err != 0)
+		return PSA_ERROR_COMMUNICATION_FAILURE;
+
+	FMSG("[RSE-COMMS] Received reply\n");
+	FMSG("protocol_ver=%u\n", io_buf.reply.header.protocol_ver);
+	FMSG("seq_num=%u\n", io_buf.reply.header.seq_num);
+	FMSG("client_id=%u\n", io_buf.reply.header.client_id);
+
+	status = rse_protocol_deserialize_reply(out_vec, out_len, &return_val,
+						&io_buf.reply, reply_size);
+	if (status != PSA_SUCCESS)
+		return status;
+
+	FMSG("return_val=%d\n", return_val);
+	for (idx = 0U; idx < out_len; idx++) {
+		FMSG("out_vec[%lu].len=%lu\n", idx, out_vec[idx].len);
+		FMSG("out_vec[%lu].buf=%p\n", idx, (void *)out_vec[idx].base);
+	}
+
+	/* Clear the mailbox message buffer to remove assets from memory */
+	memset(&io_buf, 0x0, sizeof(io_buf));
+
+	seq_num++;
+
+	return return_val;
+}
+
diff --git a/core/lib/rse_comms/src/rse_comms_protocol.c b/core/lib/rse_comms/src/rse_comms_protocol.c
new file mode 100644
index 000000000..b4ec6efd2
--- /dev/null
+++ b/core/lib/rse_comms/src/rse_comms_protocol.c
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2022, Arm Limited. All rights reserved.
+ *
+ */
+#include <assert.h>
+
+#include "rse_comms_protocol.h"
+
+psa_status_t rse_protocol_serialize_msg(psa_handle_t handle,
+					int16_t type,
+					const psa_invec *in_vec,
+					uint8_t in_len,
+					const psa_outvec *out_vec,
+					uint8_t out_len,
+					struct serialized_rse_comms_msg_t *msg,
+					size_t *msg_len)
+{
+	psa_status_t status;
+
+	assert(msg);
+	assert(msg_len);
+	assert(in_vec);
+
+	switch (msg->header.protocol_ver) {
+	case RSE_COMMS_PROTOCOL_EMBED:
+		status = rse_protocol_embed_serialize_msg(handle, type, in_vec,
+							  in_len, out_vec,
+							  out_len,
+							  &msg->msg.embed,
+							  msg_len);
+		if (status != PSA_SUCCESS)
+			return status;
+		break;
+	case RSE_COMMS_PROTOCOL_POINTER_ACCESS:
+		status = rse_protocol_pointer_access_serialize_msg(handle, type,
+								   in_vec,
+								   in_len,
+								   out_vec,
+								   out_len,
+								   &msg->msg.pointer_access,
+								   msg_len);
+		if (status != PSA_SUCCESS)
+			return status;
+		break;
+	default:
+		return PSA_ERROR_NOT_SUPPORTED;
+	}
+
+	*msg_len += sizeof(struct serialized_rse_comms_header_t);
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t rse_protocol_deserialize_reply(psa_outvec *out_vec,
+					    uint8_t out_len,
+					    psa_status_t *return_val,
+					    const struct
+					    serialized_rse_comms_reply_t *reply,
+					    size_t reply_size)
+{
+	assert(reply);
+	assert(return_val);
+
+	switch (reply->header.protocol_ver) {
+	case RSE_COMMS_PROTOCOL_EMBED:
+		return rse_protocol_embed_deserialize_reply(out_vec, out_len,
+							    return_val,
+							    &reply->reply.embed,
+							    reply_size);
+	case RSE_COMMS_PROTOCOL_POINTER_ACCESS:
+		return rse_protocol_pointer_access_deserialize_reply(out_vec,
+								     out_len,
+								     return_val,
+								     &reply->reply.pointer_access,
+								     reply_size);
+	default:
+		return PSA_ERROR_NOT_SUPPORTED;
+	}
+
+	return PSA_SUCCESS;
+}
diff --git a/core/lib/rse_comms/src/rse_comms_protocol.h b/core/lib/rse_comms/src/rse_comms_protocol.h
new file mode 100644
index 000000000..cfc2a023c
--- /dev/null
+++ b/core/lib/rse_comms/src/rse_comms_protocol.h
@@ -0,0 +1,61 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022, Arm Limited. All rights reserved.
+ *
+ */
+
+#ifndef __RSE_COMMS_PROTOCOL_H__
+#define __RSE_COMMS_PROTOCOL_H__
+
+#include <compiler.h>
+#include <psa/client.h>
+#include "rse_comms_protocol_embed.h"
+#include "rse_comms_protocol_pointer_access.h"
+#include <stdint.h>
+
+enum rse_comms_protocol_version_t {
+	RSE_COMMS_PROTOCOL_EMBED = 0,
+	RSE_COMMS_PROTOCOL_POINTER_ACCESS = 1,
+};
+
+struct __packed serialized_rse_comms_header_t {
+	uint8_t protocol_ver;
+	uint8_t seq_num;
+	uint16_t client_id;
+};
+
+/* message passed from Host to RSE to deliver a PSA client call */
+struct __packed serialized_rse_comms_msg_t {
+	struct serialized_rse_comms_header_t header;
+	union __packed {
+		struct rse_embed_msg_t embed;
+		struct rse_pointer_access_msg_t pointer_access;
+	} msg;
+};
+
+/* reply message to hold the PSA client reply result returned by RSE */
+struct __packed serialized_rse_comms_reply_t {
+	struct serialized_rse_comms_header_t header;
+	union __packed {
+		struct rse_embed_reply_t embed;
+		struct rse_pointer_access_reply_t pointer_access;
+	} reply;
+};
+
+psa_status_t rse_protocol_serialize_msg(psa_handle_t handle,
+					int16_t type,
+					const psa_invec *in_vec,
+					uint8_t in_len,
+					const psa_outvec *out_vec,
+					uint8_t out_len,
+					struct serialized_rse_comms_msg_t *msg,
+					size_t *msg_len);
+
+psa_status_t rse_protocol_deserialize_reply(psa_outvec *out_vec,
+					    uint8_t out_len,
+					    psa_status_t *return_val,
+					    const struct
+					    serialized_rse_comms_reply_t *reply,
+					    size_t reply_size);
+
+#endif /* __RSE_COMMS_PROTOCOL_H__ */
diff --git a/core/lib/rse_comms/src/rse_comms_protocol_common.h b/core/lib/rse_comms/src/rse_comms_protocol_common.h
new file mode 100644
index 000000000..6ce0b30d4
--- /dev/null
+++ b/core/lib/rse_comms/src/rse_comms_protocol_common.h
@@ -0,0 +1,34 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2024, Arm Limited. All rights reserved.
+ *
+ */
+
+/*
+ * Packing scheme of the control parameter
+ *
+ *  31           30-28   27    26-24  23-20   19     18-16   15-0
+ * +------------+-----+------+-------+-----+-------+-------+------+
+ * |            |     |      | invec |     |       | outvec| type |
+ * | Res        | Res | Res  | number| Res | Res   | number|      |
+ * +------------+-----+------+-------+-----+-------+-------+------+
+ *
+ * Res: Reserved.
+ */
+
+#ifndef RSE_COMMS_PROTOCOL_COMMON
+#define RSE_COMMS_PROTOCOL_COMMON
+
+#define TYPE_OFFSET	(0U)
+#define TYPE_MASK	(0xFFFFUL << TYPE_OFFSET)
+#define OUT_LEN_OFFSET	(16U)
+#define OUT_LEN_MASK	(0x7UL << OUT_LEN_OFFSET)
+#define IN_LEN_OFFSET	(24U)
+#define IN_LEN_MASK	(0x7UL << IN_LEN_OFFSET)
+
+#define PARAM_PACK(type, in_len, out_len)			   \
+	(((((uint32_t)(type)) << TYPE_OFFSET) & TYPE_MASK)	 | \
+	 ((((uint32_t)(in_len)) << IN_LEN_OFFSET) & IN_LEN_MASK) | \
+	 ((((uint32_t)(out_len)) << OUT_LEN_OFFSET) & OUT_LEN_MASK))
+
+#endif /* RSE_COMMS_PROTOCOL_COMMON */
diff --git a/core/lib/rse_comms/src/rse_comms_protocol_embed.c b/core/lib/rse_comms/src/rse_comms_protocol_embed.c
new file mode 100644
index 000000000..2a3acc59f
--- /dev/null
+++ b/core/lib/rse_comms/src/rse_comms_protocol_embed.c
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2022-2024, Arm Limited. All rights reserved.
+ *
+ */
+
+#include <assert.h>
+#include "rse_comms_protocol_common.h"
+#include "rse_comms_protocol_embed.h"
+#include <string.h>
+
+psa_status_t rse_protocol_embed_serialize_msg(psa_handle_t handle,
+					      int16_t type,
+					      const psa_invec *in_vec,
+					      uint8_t in_len,
+					      const psa_outvec *out_vec,
+					      uint8_t out_len,
+					      struct rse_embed_msg_t *msg,
+					      size_t *msg_len)
+{
+	uint32_t payload_size = 0;
+	uint32_t i;
+
+	assert(msg);
+	assert(msg_len);
+	assert(in_vec);
+
+	msg->ctrl_param = PARAM_PACK(type, in_len, out_len);
+	msg->handle = handle;
+
+	/* Fill msg iovec lengths */
+	for (i = 0U; i < in_len; ++i)
+		msg->io_size[i] = in_vec[i].len;
+
+	for (i = 0U; i < out_len; ++i)
+		msg->io_size[in_len + i] = out_vec[i].len;
+
+	for (i = 0U; i < in_len; ++i) {
+		if (in_vec[i].len > sizeof(msg->trailer) - payload_size)
+			return PSA_ERROR_INVALID_ARGUMENT;
+
+		memcpy(msg->trailer + payload_size,
+		       in_vec[i].base,
+		       in_vec[i].len);
+		payload_size += in_vec[i].len;
+	}
+
+	/* Output the actual size of the message, to optimize sending */
+	*msg_len = sizeof(*msg) - sizeof(msg->trailer) + payload_size;
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t rse_protocol_embed_deserialize_reply(psa_outvec *out_vec,
+						  uint8_t out_len,
+						  psa_status_t *return_val,
+						  const struct rse_embed_reply_t
+						  *reply,
+						  size_t reply_size)
+{
+	uint32_t payload_offset = 0;
+	uint32_t i;
+
+	assert(reply);
+	assert(return_val);
+
+	for (i = 0U; i < out_len; ++i) {
+		if ((sizeof(*reply) - sizeof(reply->trailer) + payload_offset)
+		    > reply_size) {
+			return PSA_ERROR_INVALID_ARGUMENT;
+		}
+
+		memcpy(out_vec[i].base,
+		       reply->trailer + payload_offset,
+		       reply->out_size[i]);
+		out_vec[i].len = reply->out_size[i];
+		payload_offset += reply->out_size[i];
+	}
+
+	*return_val = reply->return_val;
+
+	return PSA_SUCCESS;
+}
diff --git a/core/lib/rse_comms/src/rse_comms_protocol_embed.h b/core/lib/rse_comms/src/rse_comms_protocol_embed.h
new file mode 100644
index 000000000..ee65a2210
--- /dev/null
+++ b/core/lib/rse_comms/src/rse_comms_protocol_embed.h
@@ -0,0 +1,45 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022, Arm Limited. All rights reserved.
+ *
+ */
+
+#ifndef __RSE_COMMS_PROTOCOL_EMBED_H__
+#define __RSE_COMMS_PROTOCOL_EMBED_H__
+
+#include <compiler.h>
+
+#include <psa/client.h>
+
+#include <platform_config.h>
+
+struct __packed rse_embed_msg_t {
+	psa_handle_t handle;
+	uint32_t ctrl_param; /* type, in_len, out_len */
+	uint16_t io_size[PSA_MAX_IOVEC];
+	uint8_t trailer[PLAT_RSE_COMMS_PAYLOAD_MAX_SIZE];
+};
+
+struct __packed rse_embed_reply_t {
+	int32_t return_val;
+	uint16_t out_size[PSA_MAX_IOVEC];
+	uint8_t trailer[PLAT_RSE_COMMS_PAYLOAD_MAX_SIZE];
+};
+
+psa_status_t rse_protocol_embed_serialize_msg(psa_handle_t handle,
+					      int16_t type,
+					      const psa_invec *in_vec,
+					      uint8_t in_len,
+					      const psa_outvec *out_vec,
+					      uint8_t out_len,
+					      struct rse_embed_msg_t *msg,
+					      size_t *msg_len);
+
+psa_status_t rse_protocol_embed_deserialize_reply(psa_outvec *out_vec,
+						  uint8_t out_len,
+						  psa_status_t *return_val,
+						  const struct rse_embed_reply_t
+						  *reply,
+						  size_t reply_size);
+
+#endif /* __RSE_COMMS_PROTOCOL_EMBED_H__ */
diff --git a/core/lib/rse_comms/src/rse_comms_protocol_pointer_access.c b/core/lib/rse_comms/src/rse_comms_protocol_pointer_access.c
new file mode 100644
index 000000000..e53b714aa
--- /dev/null
+++ b/core/lib/rse_comms/src/rse_comms_protocol_pointer_access.c
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2022-2024, Arm Limited. All rights reserved.
+ *
+ */
+#include <assert.h>
+
+#include "rse_comms_protocol_common.h"
+#include "rse_comms_protocol_pointer_access.h"
+
+psa_status_t rse_protocol_pointer_access_serialize_msg(psa_handle_t handle,
+						       int16_t type,
+						       const psa_invec *in_vec,
+						       uint8_t in_len,
+						       const psa_outvec
+						       *out_vec,
+						       uint8_t out_len,
+						       struct
+						       rse_pointer_access_msg_t
+						       *msg,
+						       size_t *msg_len)
+{
+	unsigned int i;
+
+	assert(msg);
+	assert(msg_len);
+	assert(in_vec);
+
+	msg->ctrl_param = PARAM_PACK(type, in_len, out_len);
+	msg->handle = handle;
+
+	/* Fill msg iovec lengths */
+	for (i = 0U; i < in_len; ++i) {
+		msg->io_sizes[i] = in_vec[i].len;
+		msg->host_ptrs[i] = (uint64_t)in_vec[i].base;
+	}
+	for (i = 0U; i < out_len; ++i) {
+		msg->io_sizes[in_len + i] = out_vec[i].len;
+		msg->host_ptrs[in_len + i] = (uint64_t)out_vec[i].base;
+	}
+
+	*msg_len = sizeof(*msg);
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t rse_protocol_pointer_access_deserialize_reply(psa_outvec *out_vec,
+							   uint8_t out_len,
+							   psa_status_t
+							   *return_val,
+							   const struct
+							   rse_pointer_access_reply_t
+							   *reply,
+							   size_t reply_size)
+{
+	unsigned int i;
+
+	assert(reply);
+	assert(return_val);
+	/*  clean response len */
+	for (i = 0U; i < out_len; ++i)
+		out_vec[i].len = 0;
+	/*  compute the len that can be filled */
+	reply_size /= 4;
+	out_len = reply_size < out_len ? out_len : reply_size;
+	for (i = 0U; i < out_len; ++i)
+		out_vec[i].len = reply->out_sizes[i];
+
+	*return_val = reply->return_val;
+
+	return PSA_SUCCESS;
+}
diff --git a/core/lib/rse_comms/src/rse_comms_protocol_pointer_access.h b/core/lib/rse_comms/src/rse_comms_protocol_pointer_access.h
new file mode 100644
index 000000000..4b533756b
--- /dev/null
+++ b/core/lib/rse_comms/src/rse_comms_protocol_pointer_access.h
@@ -0,0 +1,47 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022, Arm Limited. All rights reserved.
+ *
+ */
+
+#ifndef __RSE_COMMS_PROTOCOL_POINTER_ACCESS_H__
+#define __RSE_COMMS_PROTOCOL_POINTER_ACCESS_H__
+
+#include <compiler.h>
+
+#include <psa/client.h>
+
+struct __packed rse_pointer_access_msg_t {
+	psa_handle_t handle;
+	uint32_t ctrl_param;
+	uint32_t io_sizes[PSA_MAX_IOVEC];
+	uint64_t host_ptrs[PSA_MAX_IOVEC];
+};
+
+struct __packed rse_pointer_access_reply_t {
+	int32_t return_val;
+	uint32_t out_sizes[PSA_MAX_IOVEC];
+};
+
+psa_status_t rse_protocol_pointer_access_serialize_msg(psa_handle_t handle,
+						       int16_t type,
+						       const psa_invec *in_vec,
+						       uint8_t in_len,
+						       const psa_outvec
+						       *out_vec,
+						       uint8_t out_len,
+						       struct
+						       rse_pointer_access_msg_t
+						       *msg,
+						       size_t *msg_len);
+
+psa_status_t rse_protocol_pointer_access_deserialize_reply(psa_outvec *out_vec,
+							   uint8_t out_len,
+							   psa_status_t
+							   *return_val,
+							   const struct
+							   rse_pointer_access_reply_t
+							   *reply,
+							   size_t reply_size);
+
+#endif /* __RSE_COMMS_PROTOCOL_POINTER_ACCESS_H__ */
diff --git a/core/lib/rse_comms/src/sub.mk b/core/lib/rse_comms/src/sub.mk
new file mode 100644
index 000000000..1b0e078c8
--- /dev/null
+++ b/core/lib/rse_comms/src/sub.mk
@@ -0,0 +1,3 @@
+incdirs-y += .
+srcs-y += rse_comms.c rse_comms_protocol_embed.c rse_comms_protocol.c
+srcs-y += rse_comms_protocol_pointer_access.c
diff --git a/core/lib/rse_comms/sub.mk b/core/lib/rse_comms/sub.mk
new file mode 100644
index 000000000..25712c49d
--- /dev/null
+++ b/core/lib/rse_comms/sub.mk
@@ -0,0 +1,2 @@
+global-incdirs-y += include
+subdirs-y += src
-- 
2.47.2

